node/src/Pos/Security/Workers.hs

```diff
@@ -85,8 +87,16 @@ checkForReceivedBlocksWorkerImpl
        (WorkMode ssc ctx m)
     => SendActions m -> m ()
 checkForReceivedBlocksWorkerImpl SendActions {..} = afterDelay $ do
-    repeatOnInterval (const (sec' 4)) . recoveryCommGuard $
-        whenM (needRecovery @ctx @ssc) $ triggerRecovery enqueueMsg
+    repeatOnInterval (const (sec' 4)) $ do
+        -- TODO we should be in fact checking this predicate more often.
+        -- I suggest checking it, eg, every second. But if we succeeded
+        -- and triggered recovery, we make a longer delay, say slot duration/2,
+        -- so that we don't overload the retrieval worker.
+        need <- needRecovery @ctx @ssc
+        recoveryHappening <- recoveryInProgress
+        when (need && not recoveryHappening) $ triggerRecovery enqueueMsg
```

Comment:
I don't understand the motivation for this change. Aren't the semantics the same as before?

Reply:
No, it's not. Previously the worker wouldn't kick in if we were far away but with no recovery variable in place (that's how `recoveryCommGuard` works). Now we kick in in this condition (need + don't have variable). Anyway, this worker is optimized in master already: https://github.com/input-output-hk/cardano-sl/pull/2121

Reply:
It's my understanding that `recoveryCommGuard` runs its action only if recovery is not in progress. Yes, it's true

```Haskell
-- onIgnore just logs something
recoveryCommGuard action = ifM recoveryInProgress onIgnore action
```

In practice I believe this just checks whether some `TMVar` is full.

Previously: every 4 seconds we
- Check if recovery is on and if so do nothing (`recoveryCommGuard` just logs a message)
- Otherwise check whether we need recovery and if so, trigger it

That's to say: if we're not in recovery, we will always check whether we're in recovery and trigger it if necessary. Note that it's possible for recovery to be triggered elsewhere (block retrieval code) in the mean-time, resulting in two triggers where one would have sufficed.

Now: every 4 seconds we
- Check whether we need recovery
- Check whether recovery is in progress
- Trigger recovery if and only if we need it and are not currently doing it

But isn't that the same as before? `recoveryCommGuard` ensures that `not recoveryHappening` is true (unless you account for concurrency :) ) and the inner action checks whether `needRecovery` is true, and the trigger happens only then.

Reply:
>In practice I believe this just checks whether some TMVar is full.

No.
```
recoveryCommGuard actionName action =
    getSyncStatusK >>= \case
        SSKindaSynced -> action
        status ->
            logDebug $
            sformat ("recoveryCommGuard: we are skipping action '"%stext%
                     "', because "%build) actionName status
```

Previously: we're `inf` blocks late and not doing recovery. `recoveryCommGuard` won't let us trigger recovery. Because we're not `SSKindaSynced`.

Currently: we're `inf` blocks late and not doing recovery. `need == True`, `recoveryHappening = False`.

Reply:
Oh, I see this in the cardano-sl-1.0 tree

```Haskell
class Monad m => MonadRecoveryInfo m where
    -- | Returns if 'RecoveryHeader' is 'Just' (which is equivalent to “we're
    -- doing recovery”).
    recoveryInProgress :: m Bool

-- | This is a helper function which runs given action only if we are
-- not doing recovery at this moment.  It is useful for workers which
-- shouldn't do anything while we are not synchronized.
recoveryCommGuard
    :: (MonadRecoveryInfo m, WithLogger m)
    => m () -> m ()
recoveryCommGuard action = ifM recoveryInProgress onIgnore action
  where
    onIgnore = logDebug "recoveryCommGuard: recovery in process, skipping action"
```

Reply:
Oh right, and the instance:

```Haskell
instance
    ( Monad m, MonadIO m, MonadReader ctx m
    , HasLens RecoveryHeaderTag ctx (RecoveryHeader ssc)
    ) => MonadRecoveryInfo m where
    recoveryInProgress = do
        var <- view (lensOf @RecoveryHeaderTag)
        isJust <$> atomically (STM.tryReadTMVar var)
```

Reply:
Wow, nice catch. `cardano-sl-1.0` is too old, i might have confused implementations. Anyway, it's not harmful -- my changes introduce no semantic changes.

Reply:
I thought maybe I had checked the wrong base branch. Now it's `release-1.0.4`, and maybe the recovery comm guard has changed to what you expected?

Reply:
No, you're perfectly right. My modification in this branch doesn't make any difference. I wouldn't remove it from this PR though: it will be simpler for me to merge it into `master`, because there will be a conflict (this particular piece of code is called recoveryTriggerWorker in master) and i'll check everything twice. So, yes, you're right, but let's leave it.

Reply:
Ok